compiler/noirc_frontend/src/monomorphization/ast.rs

```diff
@@ -200,6 +203,60 @@ pub enum LValue {
 
 pub type Parameters = Vec<(LocalId, /*mutable:*/ bool, /*name:*/ String, Type)>;
 
+/// Represents how an Acir function should be inlined.
+/// This type is only relevant for ACIR functions as we do not inline any Brillig functions
+#[derive(Default, Clone, Copy, PartialEq, Eq, Debug, Hash)]
+pub enum InlineType {
+    /// The most basic entry point can expect all its functions to be inlined.
+    /// All function calls are expected to be inlined into a single ACIR.
+    #[default]
+    Inline,
+    /// Functions marked as foldable will not be inlined and compiled separately into ACIR
+    Fold,
+    /// Similar to `Fold`, these functions will not be inlined and compile separately into ACIR.
+    /// They are different from `Fold` though as they are expected to be inlined into the program
+    /// entry point before being used in the backend.
```

Comment:
Just to make sure we're on the same page here. The idea is that we'll perform ACIRgen for this function and then inline it into the main ACIR as part of the compilation process right? This seems counter intuitive for something called `#[inline(never)]`

If we keep the function as a separate ACIR then it seems like there's zero difference between this and `#[fold]`.

Reply:
In this PR there is no difference. We do not intend to inline it into the main ACIR as part of the compilation process. We will need to add some metadata about which circuits should be inlined or not for a _backend_. I have made an issue here (https://github.com/noir-lang/noir/issues/4912) and this entire epic will provide more context (https://github.com/noir-lang/noir/issues/4877)

Reply:
I like to think of it this way. `#[fold]` is a codegen attribute that dictates how it should be proven to a backend. `#[inline(never)]` is a pure code gen attribute which dictates that it should be inlined by a backend (or right before by some acvm/nargo tool).

Reply:
My thinking on this is that the backend has full control over how it handles multiple ACIR functions, a `#[fold]` attribute on a function can be entirely ignored by a backend which needs everything inlined into a single circuit. Similarly if a backend can fold together non-inlined functions more efficiently than inlining them, why should we stop it?

Reply:
> My thinking on this is that the backend has full control over how it handles multiple ACIR functions

We could have the backend do the inlining instead of something like acvm/nargo. I think providing some tool in ACVM makes this much simpler for backend integration though. This is slightly orthogonal to this PR though as it simply adds an attribute.

> a #[fold] attribute on a function can be entirely ignored by a backend which needs everything inlined into a single circuit.

I'm a little confused by this? As per https://github.com/noir-lang/noir/issues/4912 we will need some inline metadata on circuits that backends or some post compilation tool can interpret.

> Similarly if a backend can fold together non-inlined functions more efficiently than inlining them, why should we stop it?

A user may not want this. This purely aims to provide them with the option to optimize codegen of a hot function for their vanilla circuit (https://github.com/noir-lang/noir/issues/4688). 

A folding prover and decider is not the same as a regular circuit. For example, barretenberg will most likely never support verifying a folding proof in a solidity verifier. A user might just want to inlining some arithmetic function (such as poseidon in #4688) over and over in an if statement based upon witness conditions.

Reply:
In a way this function is a bit of an `#[inline(delay)]` from the perspective of the backend. But I went with the name `#[inline(never)]` as the attribute refers entirely to the codegen. I can switch the naming if we feel it is more fitting, but I feel `delay` may be more confusing to devs.

Reply:
Yeah, agreed `delay` is a more confusing.